gtk4/src/dialog.rs

```diff
@@ -49,9 +53,30 @@ impl<O: IsA<Dialog>> DialogExtManual for O {
     fn get_response_for_widget<P: IsA<Widget>>(&self, widget: &P) -> ResponseType {
         unsafe {
             from_glib(ffi::gtk_dialog_get_response_for_widget(
-                self.as_ref().to_glib_none().0,
+                AsRef::<Dialog>::as_ref(self).to_glib_none().0,
                 widget.as_ref().to_glib_none().0,
             ))
         }
     }
+
+    fn run_future<'a>(&'a self) -> Pin<Box<dyn Future<Output = ResponseType> + 'a>> {
+        Box::pin(async move {
+            let (sender, receiver) = futures_channel::oneshot::channel();
+
+            let sender = std::cell::Cell::new(Some(sender));
+
+            let response_handler = self.connect_response(move |_, response_type| {
+                if let Some(m) = sender.replace(None) {
+                    let _result = m.send(response_type);
+                }
+            });
+
+            self.show();
+
+            let result = receiver.await.unwrap_or(ResponseType::None);
+            self.disconnect(response_handler);
```

Comment:
If `receiver.await` returns an `Err`, then the signal handler was freed before the dialog got a response. In that case, `disconnect()` will probably give a critical warning so you probably only want to call it in the other case

Reply:
Fixed.